File: vi/codes/cpp/chapter_array_and_linkedlist/array.cpp
/**
 * File: array.cpp
 * Created Time: 2022-11-25
 * Author: krahets (krahets@163.com)
 */

#include "../utils/common.hpp"

/* Truy cập ngẫu nhiên vào một phần tử */
int randomAccess(int *nums, int size) {
    // Chọn ngẫu nhiên một số trong khoảng [0, size)
    int randomIndex = rand() % size;
    // Lấy và trả về phần tử ngẫu nhiên
    int randomNum = nums[randomIndex];
    return randomNum;
}

/* Mở rộng độ dài mảng */
int *extend(int *nums, int size, int enlarge) {
    // Khởi tạo một mảng có độ dài đã được mở rộng
    int *res = new int[size + enlarge];
    // Sao chép tất cả các phần tử từ mảng gốc sang mảng mới
    for (int i = 0; i < size; i++) {
        res[i] = nums[i];
    }
    // Giải phóng bộ nhớ
    delete[] nums;
    // Trả về mảng mới đã được mở rộng
    return res;
}

/* Chèn phần tử num vào vị trí index trong mảng */
void insert(int *nums, int size, int num, int index) {
    // Dịch tất cả các phần tử từ vị trí index trở về sau lùi lại một vị trí
    for (int i = size - 1; i > index; i--) {
        nums[i] = nums[i - 1];
    }
    // Gán num cho phần tử tại vị trí index
    nums[index] = num;
}

/* Xóa phần tử tại vị trí index */
void remove(int *nums, int size, int index) {
    // Dịch tất cả các phần tử sau vị trí index tiến lên một vị trí
    for (int i = index; i < size - 1; i++) {
        nums[i] = nums[i + 1];
    }
}

/* Duyệt mảng */
void traverse(int *nums, int size) {
    int count = 0;
    // Duyệt mảng theo chỉ số
    for (int i = 0; i < size; i++) {
        count += nums[i];
    }
}

/* Tìm phần tử được chỉ định trong mảng */
int find(int *nums, int size, int target) {
    for (int i = 0; i < size; i++) {
        if (nums[i] == target)
            return i;
    }
    return -1;
}

/* Mã chạy chính (Driver Code) */
int main() {
    /* Khởi tạo mảng */
    int size = 5;
    int *arr = new int[size];
    cout << "Array arr = ";
    printArray(arr, size);

    int *nums = new int[size]{1, 3, 2, 5, 4};
    cout << "Array nums = ";
    printArray(nums, size);

    /* Truy cập ngẫu nhiên */
    int randomNum = randomAccess(nums, size);
    cout << "Get random element in nums " << randomNum << endl;

    /* Mở rộng độ dài mảng */
    int enlarge = 3;
    nums = extend(nums, size, enlarge);
    size += enlarge;
    cout << "Extend array length to 8, resulting in nums = ";
    printArray(nums, size);

    /* Chèn phần tử */
    insert(nums, size, 6, 3);
    cout << "Insert number 6 at index 3, resulting in nums = ";
    printArray(nums, size);

    /* Xóa phần tử */
    remove(nums, size, 2);
    cout << "Remove element at index 2, resulting in nums = ";
    printArray(nums, size);

    /* Duyệt mảng */
    traverse(nums, size);

    /* Tìm phần tử */
    int index = find(nums, size, 3);
    cout << "Find element 3 in nums, get index = " << index << endl;

    // Giải phóng bộ nhớ
    delete[] arr;
    delete[] nums;

    return 0;
}

File: vi/codes/rust/chapter_array_and_linkedlist/array.rs
/*
 * File: array.rs
 * Created Time: 2023-01-15
 * Author: xBLACICEx (xBLACKICEx@outlook.com), codingonion (coderonion@gmail.com)
 */

use hello_algo_rust::include::print_util;
use rand::Rng;

/* Truy cập ngẫu nhiên một phần tử */
fn random_access(nums: &[i32]) -> i32 {
    // Chọn ngẫu nhiên một số trong khoảng [0, nums.len())
    let random_index = rand::thread_rng().gen_range(0..nums.len());
    // Lấy và trả về phần tử ngẫu nhiên
    let random_num = nums[random_index];
    random_num
}

/* Mở rộng độ dài mảng */
fn extend(nums: &[i32], enlarge: usize) -> Vec<i32> {
    // Khởi tạo một mảng với độ dài đã mở rộng
    let mut res: Vec<i32> = vec![0; nums.len() + enlarge];
    // Sao chép toàn bộ phần tử từ mảng gốc sang mảng mới
    res[0..nums.len()].copy_from_slice(nums);

    // Trả về mảng mới đã được mở rộng
    res
}

/* Chèn phần tử num vào vị trí index trong mảng */
fn insert(nums: &mut [i32], num: i32, index: usize) {
    // Dịch chuyển tất cả phần tử từ vị trí index trở về sau lùi lại một vị trí
    for i in (index + 1..nums.len()).rev() {
        nums[i] = nums[i - 1];
    }
    // Gán num vào phần tử tại vị trí index
    nums[index] = num;
}

/* Xóa phần tử tại vị trí index */
fn remove(nums: &mut [i32], index: usize) {
    // Dịch chuyển tất cả phần tử sau vị trí index tiến lên một vị trí
    for i in index..nums.len() - 1 {
        nums[i] = nums[i + 1];
    }
}

/* Duyệt mảng */
fn traverse(nums: &[i32]) {
    let mut _count = 0;
    // Duyệt mảng theo chỉ số
    for i in 0..nums.len() {
        _count += nums[i];
    }
    // Duyệt trực tiếp các phần tử của mảng
    _count = 0;
    for &num in nums {
        _count += num;
    }
}

/* Tìm phần tử chỉ định trong mảng */
fn find(nums: &[i32], target: i32) -> Option<usize> {
    for i in 0..nums.len() {
        if nums[i] == target {
            return Some(i);
        }
    }
    None
}

/* Mã điều khiển chính */
fn main() {
    /* Khởi tạo mảng */
    let arr: [i32; 5] = [0; 5];
    print!("Array arr = ");
    print_util::print_array(&arr);
    // Trong Rust, khi chỉ định độ dài ([i32; 5]) thì đó là mảng (array), còn không chỉ định độ dài (&[i32]) thì đó là lát cắt (slice)
    // Vì mảng trong Rust được thiết kế có độ dài xác định tại thời điểm biên dịch, nên chỉ hằng số mới có thể chỉ định độ dài
    // Vector là kiểu dữ liệu mà Rust thường dùng để biểu diễn mảng động
    // Để thuận tiện cho việc triển khai phương thức extend(), phần dưới đây sẽ coi vector như mảng
    let nums: Vec<i32> = vec![1, 3, 2, 5, 4];
    print!("\nArray nums = ");
    print_util::print_array(&nums);

    // Chèn phần tử
    let random_num = random_access(&nums);
    println!("\nGet random element {} from nums", random_num);

    // Duyệt mảng
    let mut nums: Vec<i32> = extend(&nums, 3);
    print!("Extend array length to 8, resulting in nums = ");
    print_util::print_array(&nums);

    // Chèn phần tử
    insert(&mut nums, 6, 3);
    print!("\nInsert number 6 at index 3, get nums = ");
    print_util::print_array(&nums);

    // Xóa phần tử
    remove(&mut nums, 2);
    print!("\nDelete element at index 2, get nums = ");
    print_util::print_array(&nums);

    // Duyệt mảng
    traverse(&nums);

    // Tìm phần tử
    let index = find(&nums, 3).unwrap();
    println!("\nFind element 3 in nums, index = {}", index);
}
